guides/schema/class_based_api.md

```diff
@@ -0,0 +1,353 @@
+    - ☐ Custom directives
+    - ☐ Custom `Schema#execute` method
+  - ☐ Migrate all of GitHub's GraphQL schema to this new API
+- graphql 1.9:
+  - ☐ Update all GraphQL-Ruby docs to reflect this new API
+- graphql 1.10:
+  - ☐ Begin sunsetting `.define`: isolate it in its own module
+  - ☐ Remove `.define`
+
+## Schema class
+
+Your GraphQL schema is a class that extends {{ "GraphQL::Schema" | api_doc }}. Its configuration options are similar to `.define`-based options, but if you find something that doesn't work, please {% open_an_issue "Class-based schema issue","(Please share some example code and the error you found)" %}.
+
+```ruby
+class MyAppSchema < GraphQL::Schema
+  max_complexity 400
+  query Types::Query
+  use GraphQL::Batch
+
+  # Define hooks as class methods:
+  def self.resolve_type(type, obj, ctx)
+    # ...
+  end
+
+  def self.object_from_id(node_id, ctx)
+    # ...
+  end
+
+  def self.id_from_object(object, type, ctx)
+    # ...
+  end
+end
+```
+
+## Common Type Configurations
+
+Some configurations are used for _all_ types described below:
+
+- `graphql_name` overrides the type name. (The default value is the Ruby constant name, without any namespaces)
+- `description` provides a description for GraphQL introspection.
+
+For example:
+
+```ruby
+class Types::TodoList < GraphQL::Schema::Object # or Scalar, Enum, Union, whatever
+  graphql_name "List" # Overrides the default of "TodoList"
+  description "Things to do (may have already been done)"
+end
+```
+
+(Implemented in {{ "GraphQL::Schema::Member" | api_doc }}).
+
+## Object classes
+
+Classes extending {{ "GraphQL::Schema::Object" | api_doc }} describe [Object types](http://graphql.org/learn/schema/#object-types-and-fields) and customize their behavior.
+
+Object fields can be created with the `field(...)` class method, which accepts the similar arguments as the previous `field(...)` method.
+
+```ruby
+# first, somewhere, a base class:
+class Types::BaseObject < GraphQL::Schema::Object
+end
+
+# then...
+class Types::TodoList < BaseObject
+  field :name, String, "The unique name of this list", null: false
+  # Related Object:
+  field :owner, Types::User, "The creator of this list", null: false
+  # List field:
+  field
+  # Connection:
+  field :items, Types::TodoItem.connection_type, "Tasks on this list", null: false do
+    argument :status, TodoStatus, "Restrict items to this status", null: true
+  end
+end
+```
+
+### New return type & argument type specification
+
+The second argument to `field(...)` is the return type. This can be:
+
+- A GraphQL type object built with `.define { ... }`
+- A GraphQL type class which you defined
+- A Ruby built-in such as `Integer`, `Float` or `String`
+- A string which corresponds to a GraphQL built-in, such as `"ID"`, `"Boolean"`, or `"Int"`
+- An _array_ of any of the above, which denotes a list type. Inner list types are always made non-null.
+
+Nullability is expressed with the required `null:` keyword:
```

Comment:
Hmmm, this syntax is definitely more Rubyish than the `!` overloading, but not having a default of non-nullable seems counter to the GraphQL spec to me. `String` is nullable in the spec, and IMO it should be in the Ruby DSL too. 

In other words, I think these should be valid and equivalent:
```ruby
field :name, String, null: true
field :name, String
```

Reply:
I argued for making it required somewhere up above. What we've found internally is that if we don't make it required developers forget to think about it entirely and end up making fields nullable that should really be non-null.